### PR TITLE
Animation Blending

### PR DESCRIPTION
**Pull request by Bendzae**
# Objective

- Extend the bevy animation system to allow blending between animations
  - Should be able to specify the duration of the transition
  - Both the current and the target animations should run while blending between them
  - Provide a working example

## Solution

Extended the `AnimationPlayer` with a function: 
`pub fn cross_fade(&mut self, handle: Handle<AnimationClip>, transition_time: f32) -> &mut AnimationTransition`
This initiates a fade from the current to the provided animation.

The information about this "Transition" is stored in a the struct `AnimationTransition` and is added to the `AnimationPlayer`. 

The `animation_player` system was adapted to handle this new data by interpolating the transforms of the targeted entities between the transforms of the keyframes of the current and target clip.

---

### Added
- `AnimationTransition`: Representation of the data needed to transition from one to another animation 

### Changed
- `AnimationPlayer`
  - added the `cross_fade()` function 
  - added the `transition` field
- `animation_player` system
- `animated_fox.rs`: Adapted example to be able to crossfade or normale play animations

### CODE REVIEW

File: crates/bevy_animation/src/lib.rs [hunk 1]
```diff
@@ -224,61 +330,112 @@ pub fn animation_player(
...
+                            match &curve.keyframes {
+                                Keyframes::Rotation(keyframes) => {
+                                    let rot_start = keyframes[step_start];
+                                    let mut rot_end = keyframes[step_start + 1];
+                                    // Choose the smallest angle for the rotation
+                                    if rot_end.dot(rot_start) < 0.0 {
+                                        rot_end = -rot_end;
+                                    }
+
+                                    // Rotations are using a spherical linear interpolation
```
Comment (alice-i-cecile) [inline crates/bevy_animation/src/lib.rs:392]: Excellent, this is one of the things I was looking for.

File: crates/bevy_animation/src/lib.rs [hunk 2]
```diff
@@ -224,61 +330,112 @@ pub fn animation_player(
...
+                        transform.rotation = from.rotation.slerp(to.rotation, transition_lerp);
+                        transform.translation =
+                            from.translation.lerp(to.translation, transition_lerp);
+                        transform.scale = from.scale.lerp(to.scale, transition_lerp);
+                    }
+                }
+            }
+
+            // Transition to next clip has finished
+            if transition_lerp == 1.0 {
```
Comment (alice-i-cecile) [inline crates/bevy_animation/src/lib.rs:429]: I'm nervous about exact float equality. I'd prefer a >= here

File: crates/bevy_animation/src/lib.rs [hunk 3]
```diff
@@ -90,6 +90,58 @@ impl AnimationClip {
     }
 }
 
+/// A representation of the data needed transition to a new animation
```
Comment (alice-i-cecile) [inline crates/bevy_animation/src/lib.rs:93]:
```suggestion
/// The data needed transition to a new animation
///
/// Stored in an [`AnimationPlayer`]
```

File: crates/bevy_animation/src/lib.rs [hunk 4]
```diff
@@ -90,6 +90,58 @@ impl AnimationClip {
...
+#[derive(Clone, Debug, Reflect)]
+pub struct AnimationTransition {
+    next_animation_clip: Handle<AnimationClip>,
+    repeat: bool,
+    speed: f32,
+    /// The elapsed time of the target clip (is scaled by speed and reset on clip end)
+    clip_elapsed: f32,
```
Comment (alice-i-cecile) [inline crates/bevy_animation/src/lib.rs:100]: We should be using `Duration` to represent periods of time, not raw f32s, unless we have a very good reason not to.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]

**Comment by alice-i-cecile** [comment]
Surprisingly straightforward. I'm not entirely sure how this fits into the broader plans in this area, but given the amount of added functionality relative to the complexity and the importance of Transform animation (which is likely going to end up special-cased) I'm inclined to merge this before the RFCs are in.

File: crates/bevy_animation/src/lib.rs (hunk 3 above)
Comment (Bendzae) [reply crates/bevy_animation/src/lib.rs]: done

File: crates/bevy_animation/src/lib.rs (hunk 4 above)
Comment (Bendzae) [reply crates/bevy_animation/src/lib.rs]: Now using Duration for all fields that represent a period of time.

File: crates/bevy_animation/src/lib.rs (hunk 2 above)
Comment (Bendzae) [reply crates/bevy_animation/src/lib.rs]: fixed :)

**Review by alice-i-cecile — APPROVED** [review]
LGTM now :) You can resolve all of my comments.

**Review by james7132 — COMMENTED** [review]
While useful in its own right, this implementation is a bit shortsighted. This only supports a single linear transition between two clips, whereas the animation RFCs aim to support an arbitrary N-wise linear blend of multiple clips. This not only supports a transition between single clips, but also more complex blending structures like state machines, layered blends, or N-D blend trees.

Overall, I think we should prioritize reviewing and implementing the RFCs over extending the current animation player, which we've acknowledged was meant as a short term stopgap to ship with 0.7.

**Comment by Bendzae** [comment]
> While useful in its own right, this implementation is a bit shortsighted. This only supports a single linear transition between two clips, whereas the animation RFCs aim to support an arbitrary N-wise linear blend of multiple clips. This not only supports a transition between single clips, but also more complex blending structures like state machines, layered blends, or N-D blend trees.
> 
> Overall, I think we should prioritize reviewing and implementing the RFCs over extending the current animation player, which we've acknowledged was meant as a short term stopgap to ship with 0.7.

I totally agree. My line of thought was just to put it out there to provide an intermediary solution until the RFCs are implemented. So people like myself, that need at least some basic form of animation blending for their project, have something to work with. 

If you think it will somehow hinder the progress or is too much effort to review I would have no Problem with closing this PR for now.

**Comment by mockersf** [comment]
This increase the system `animation_player` duration by 1.4ms in many_foxes examples, without any transitions between animations
Screenshot 2022-07-23 at 13 01 14
red is main, yellow is this PR

Would it be possible to reduce the impact when there are no transitions ongoing?

**Comment by Bendzae** [comment]
> 

Hi Mockersf,
Sorry for the late reply, what profiling tool are you using to generate that graph?
I will try to optimize frametimes for non-transition animations asap.

**Comment by mockersf** [comment]
I am using Tracy: https://github.com/wolfpld/tracy.
There are some docs in how to use with Bevy here:  https://github.com/bevyengine/bevy/blob/main/docs/profiling.md#backend-trace_tracy

You'll need to enable Bevy feature `trace_tracy`: `cargo run --release --example many_foxes --features trace_tracy`

**Comment by Bendzae** [comment]
> This increase the system `animation_player` duration by 1.4ms in many_foxes examples, without any transitions between animations crossfade red is main, yellow is this PR
> 
> Would it be possible to reduce the impact when there are no transitions ongoing?

I managed to get it down to a difference of about .1 ms on my machine. Would that be fine with you @mockersf ?

image

**Comment by mockersf** [comment]
Animation changed since this PR, and this feature should be covered by #6922

**Pull request closed without merging**